fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingThrottle.java

```diff
@@ -385,26 +385,34 @@ public <R> CompletableFuture<R> buildCommitRetryAsync(@Nonnull final BiFunction<
 
     private void expectedIndexStatesOrThrow(FDBRecordStore store, FDBRecordContext context) {
         List<IndexState> indexStates = common.getTargetIndexes().stream().map(store::getIndexState).collect(Collectors.toList());
-        if (indexStates.stream().allMatch(state -> state == expectedIndexState)) {
+        if (isScrubber) {
+            // index scrubbing requires a scannable state
+            if (indexStates.stream().allMatch(IndexState::isScannable)) {
+                return;
+            }
+            throw new IndexingBase.UnexpectedReadableException(false, "Attempt to scrub a non readable index",
+                    LogMessageKeys.INDEX_STATE, indexStates);
+        }
+        // Here: index building
+        if (indexStates.stream().allMatch(IndexState::isWriteOnly)) {
             return;
         }
         // possible exceptions:
         // 1. All the indexes are now readable.
         // 2. Some indexes are built, but all the others are in the expected state.
         // 3. Some indexes are not in the expected state (disabled?).
         // During mutual indexing, the first two may be part of the valid path
-        if (indexStates.stream().allMatch(state -> (state == IndexState.READABLE || state == IndexState.READABLE_UNIQUE_PENDING))) {
+        if (indexStates.stream().allMatch(IndexState::isScannable)) {
```

Comment:
Does the transition from `READABLE_UNIQUE_PENDING` to `READABLE` happen before this code, such that if you try to build indexes that are `READABLE_UNIQUE_PENDING` it will succeed? `testReadableUniquePendingSingle` implies that it should work.

Reply:
Yes. 
The `DesiredAction` for indexes in this state is `DesiredAction.MARK_READABLE`. This was designed to save resources when an index with uniqueness violation is attempted to be built.